include/exadg/postprocessor/time_control_statistics.cpp

```diff
@@ -0,0 +1,85 @@
+/*  ______________________________________________________________________
+ *
+ *  ExaDG - High-Order Discontinuous Galerkin for the Exa-Scale
+ *
+ *  Copyright (C) 2021 by the ExaDG authors
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+// ExaDG
+#include <exadg/postprocessor/time_control_statistics.h>
+#include <exadg/utilities/print_functions.h>
+
+namespace ExaDG
+{
+TimeControlDataStatistics::TimeControlDataStatistics()
+  : write_preliminary_results_every_nth_time_step(numbers::invalid_timestep)
+{
+}
+
+void
+TimeControlDataStatistics::print(dealii::ConditionalOStream & pcout, bool const unsteady) const
+{
+  time_control_data.print(pcout, unsteady);
+  if(Utilities::is_valid_timestep(write_preliminary_results_every_nth_time_step))
+    print_parameter(pcout,
+                    "Write preliminary results every nth time step",
+                    write_preliminary_results_every_nth_time_step);
+}
+
+TimeControlStatistics::TimeControlStatistics() : final_output_written(false)
+{
+}
+
+void
+TimeControlStatistics::setup(TimeControlDataStatistics const & time_control_data_statistics_in)
+{
+  time_control_data_statistics = time_control_data_statistics_in;
+  time_control.setup(time_control_data_statistics.time_control_data);
+}
+
+bool
+TimeControlStatistics::write_preliminary_results(double const           time,
+                                                 types::time_step const time_step_number) const
+{
+  if(Utilities::is_valid_timestep(
+       time_control_data_statistics.write_preliminary_results_every_nth_time_step) &&
+     !final_output_written)
+  {
+    if(time <
+       time_control_data_statistics.time_control_data.start_time - time_control.get_epsilon())
+    {
+      return false;
+    }
+
+    if(time_control.reached_end_time())
+    {
+      final_output_written = true;
+      return true;
+    }
+
+    if((time_step_number - 1) %
+         (time_control_data_statistics.write_preliminary_results_every_nth_time_step) ==
+       0)
+    {
+      return true;
+    }
+  }
```

Comment:
I want to make a suggestion below how to write this code more elegantly and safer in my opinion. I do not like so much the style with many returns inside a function. Why? Changing the order of the last two if-statements will in general change the logic of the code (!), but this is very hard to see and to understand for someone that has not written the code. Hence, this bug might easily be introduced in the future during refactoring.

```c++
  if(Utilities::is_valid_timestep(
       time_control_data_statistics.write_preliminary_results_every_nth_time_step))
  {
    // if we are outside the interval bounds
    if((time <
       time_control_data_statistics.time_control_data.start_time - time_control.get_epsilon())
      or final_output_written)
    {
      return false;
     }
    // check whether results have to be written
    else if((time_step_number - 1) %
         (time_control_data_statistics.write_preliminary_results_every_nth_time_step) ==
       0 or time_control.reached_end_time())
    {
      // make sure that we do no longer write results once the final output has been written
      if(time_control.reached_end_time())
        final_output_written = true;

      return true;
    }
  }
```

Reply:
Im OK with both versions